### PR TITLE
Various scheme-related optimizations.

### PR DESCRIPTION
**Pull request by lemire**
The main trick that this is using is that instead of, for example, passing the view "http" and then later creating the temporary "http:" which gets added to our buffer, we just pass straight up  "http:" and avoid the messy business entirely. Also, I created new methods to set the URL to file type, very fast (instead of passing the unnecessary string "file").

Apple M2, LLVM 14, benchdata.

Before:
```
Benchmark                                        Time             CPU   Iterations        GHz cycle/byte cycles/url instructions/byte instructions/cycle instructions/ns instructions/url     ns/url      speed  time/byte   time/url      url/s
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

BasicBench_AdaURL_aggregator              25379323 ns     25377464 ns           28    3.40142    9.73549    845.617            42.563            4.37194         14.8708         3.69699k    248.607 342.355M/s  2.92095ns  253.711ns 3.94149M/s
BasicBench_AdaURL_aggregator_just_parse   25524323 ns     25523074 ns           27    3.32936    9.76305    848.011            42.482            4.35131         14.4871         3.68996k    254.707 340.401M/s  2.93771ns  255.167ns   3.919M/s
```

After:
```
Benchmark                                        Time             CPU   Iterations        GHz cycle/byte cycles/url instructions/byte instructions/cycle instructions/ns instructions/url     ns/url      speed  time/byte   time/url      url/s
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
BasicBench_AdaURL_aggregator              22431659 ns     22431677 ns           31    3.39537    8.73146    758.408           39.4446            4.51752         15.3387         3.42613k    223.365 387.314M/s  2.58189ns  224.261ns  4.4591M/s
BasicBench_AdaURL_aggregator_just_parse   22537055 ns     22537065 ns           31    3.40045    8.76141    761.009           39.4039            4.49744         15.2933         3.42259k    223.797 385.502M/s  2.59402ns  225.314ns 4.43824M/s
```

### CODE REVIEW

File: src/url_aggregator.cpp [hunk 1]
```diff
@@ -107,22 +183,46 @@ inline void url_aggregator::copy_scheme(const url_aggregator& u) noexcept {
   ADA_ASSERT_TRUE(validate());
 }
 
+inline void url_aggregator::set_scheme_fast(std::string_view new_scheme_with_colon) noexcept {
```
Comment (anonrig) [inline src/url_aggregator.cpp:186]: Now I see that, this is a function for having an input with a colon. I think the function name should reflect the functionality.

File: src/url_aggregator.cpp [hunk 2]
```diff
@@ -107,22 +183,46 @@ inline void url_aggregator::copy_scheme(const url_aggregator& u) noexcept {
...
+  ADA_ASSERT_TRUE(!new_scheme_with_colon.empty() && new_scheme_with_colon.back() == ':');
+  uint32_t new_difference = uint32_t(new_scheme_with_colon.size()) - components.protocol_end;
+
+  if(buffer.empty()) {
+    buffer.append(new_scheme_with_colon);
+  } else {
+    buffer.erase(0, components.protocol_end);
+    buffer.insert(0, new_scheme_with_colon);
+  }
+  components.protocol_end = uint32_t(new_scheme_with_colon.size());
```
Comment (anonrig) [inline src/url_aggregator.cpp:198]: This can be reduced to 

```
components.protocol_end += new_difference;
```

File: src/url_aggregator.cpp [hunk 3]
```diff
@@ -107,22 +183,46 @@ inline void url_aggregator::copy_scheme(const url_aggregator& u) noexcept {
   ADA_ASSERT_TRUE(validate());
 }
 
+inline void url_aggregator::set_scheme_fast(std::string_view new_scheme_with_colon) noexcept {
+  ada_log("url_aggregator::set_scheme_fast ", new_scheme_with_colon);
+  ADA_ASSERT_TRUE(validate());
+  ADA_ASSERT_TRUE(!new_scheme_with_colon.empty() && new_scheme_with_colon.back() == ':');
+  uint32_t new_difference = uint32_t(new_scheme_with_colon.size()) - components.protocol_end;
```
Comment (anonrig) [inline src/url_aggregator.cpp:190]: If you have a `https` and later change your scheme to `ftp`, due to unsigned integer constraint, we would have a really large unsigned integer, right? This applies to every function in the url_aggregator

Comment (lemire) [reply src/url_aggregator.cpp:190]: Yes but it does not matter. Unsigned overflow is well defined in C++

Comment (anonrig) [reply src/url_aggregator.cpp:190]: Wouldn't new_difference be too big of a number, effecting the updates of each url_components? How does that not create a bug. I think I'm missing something.

File: include/ada/url_aggregator.h [hunk 4]
```diff
@@ -42,12 +42,17 @@ namespace ada {
     void set_search(const std::string_view input);
     void set_hash(const std::string_view input);
     inline void set_scheme(std::string_view new_scheme) noexcept;
+    /** @private */
+    inline void set_scheme_fast(std::string_view new_scheme_with_colon) noexcept;
```
[the remark this answers is not in the document]
Comment (lemire) [reply include/ada/url_aggregator.h:46]: Done.

File: src/parser.cpp [hunk 5]
```diff
@@ -659,8 +664,7 @@ namespace ada::parser {
           ada_log("FILE ", helpers::substring(url_data, input_position));
           std::string_view file_view = helpers::substring(url_data, input_position);
 
-          // Set url’s scheme to "file".
-          url.set_scheme("file");
+          url.set_file_protocol();
```
[the remark this answers is not in the document]
Comment (lemire) [reply src/parser.cpp:667]: Done.

File: src/url_aggregator.cpp [hunk 6]
```diff
@@ -69,7 +69,83 @@ template <bool has_state_override>
...
     }
 
-    set_scheme(std::move(_buffer));
+    set_scheme(_buffer);
+
+    if (has_state_override) {
+      // This is uncommon.
+      uint16_t urls_scheme_port = get_special_port();
+
+      if (urls_scheme_port) {
```
[the remark this answers is not in the document]
Comment (lemire) [reply src/url_aggregator.cpp:78]: I did not change this code, but sure...

[the remark this answers is not in the document]
Comment (lemire) [reply src/url_aggregator.cpp:78]: Done.

File: src/url_aggregator.cpp [hunk 7]
```diff
@@ -69,7 +69,83 @@ template <bool has_state_override>
...
+  ADA_ASSERT_TRUE(!helpers::overlaps(input_with_colon, buffer));
+  std::string_view input{input_with_colon};
+  input.remove_suffix(1);
+  auto parsed_type = ada::scheme::get_scheme_type(input);
+  bool is_input_special = (parsed_type != ada::scheme::NOT_SPECIAL);
+  /**
+   * In the common case, we will immediately recognize a special scheme (e.g., http, https),
+   * in which case, we can go really fast.
+   **/
+  if(is_input_special) { // fast path!!!
```
[the remark this answers is not in the document]
Comment (lemire) [reply src/url_aggregator.cpp:102]: A fair concern. I have removed the older parsing function.

File: src/url_aggregator.cpp (hunk 1 above)
Comment (lemire) [reply src/url_aggregator.cpp:186]: Done.

File: src/url_aggregator.cpp (hunk 2 above)
Comment (lemire) [reply src/url_aggregator.cpp:198]: Change made.

File: src/url_aggregator.cpp (hunk 3 above)
Comment (lemire) [reply src/url_aggregator.cpp:190]: What is `12 + 4294967295` as `uint32_t`? It is 11.

Comment (lemire) [reply src/url_aggregator.cpp:190]: I will add a comment. If you think there is a bug, consider adding a test case. I don't think there is a bug.

Comment (lemire) [reply src/url_aggregator.cpp:190]: Note also that it is not new to this PR.

**Pull request merged**